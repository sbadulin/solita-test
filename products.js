/**
 * Mock API served as Express middleware
 */
module.exports = {
  products: [
    {
      id: 1,
      name: "Circle",
      description: "Perfect choice when you don’t need any corners.",
      price: 999
    },
    {
      id: 2,
      name: "Rectangle",
      description: "For once, it’s a great idea to think inside the box.",
      price: 899
    },
    {
      id: 3,
      name: "Triangle",
      description: "A true classic with three elegant corners.",
      price: 1009
    }
  ]
};
