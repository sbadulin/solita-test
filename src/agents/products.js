import requests from "./helpers/requests";

export default {
  get: productId =>
    productId
      ? requests.get(`/products/${productId}`)
      : requests.get(`/products`)
};
