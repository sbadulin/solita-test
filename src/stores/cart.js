/**
 * Cart store
 *
 * Stores product ids in cart, cart size, removing and adding methods
 */
import { action, observable, computed } from "mobx";

class Cart {
  @observable cartItemIds = [];

  /**
   * Counts items in cart
   *
   * @returns {number} number of items in cart
   */

  @computed
  get cartSize() {
    return this.cartItemIds.length;
  }

  /**
   * Adds product id to cartItemIds
   *
   * @param {number} productId
   * @returns {null}
   */
  @action
  addToCart = productId => {
    this.cartItemIds.push(productId);
  };

  /**
   * Removes product id from cartItemIds
   *
   * @param {number} productId
   * @returns {null}
   */
  @action
  removeFromCart = productId => {
    const index = this.cartItemIds.indexOf(productId);
    this.cartItemIds.splice(index, 1);
  };
}

export default new Cart();
