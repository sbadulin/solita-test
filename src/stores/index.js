export { default as productsStore } from "./products";
export { default as cartStore } from "./cart";
