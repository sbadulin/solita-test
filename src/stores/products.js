/**
 * Products store
 *
 * Gets products from a server
 * Stores gotten products in registry
 */
import { action, flow, observable } from "mobx";
import { normalize } from "normalizr";
import { productsAgent } from "../agents";
import { updateItemsInRegistry } from "./helpers/registry";
import { productSchema } from "../schemas";

export class Products {
  @observable productLoading = false;

  @observable registry = {};

  /**
   * Gets products from a server
   *
   * @async
   * @param {null} no params
   * @returns {Promise<Array<Item>>} promise with array of products
   */
  @action
  getProducts = flow(function* getProducts() {
    try {
      const { products } = yield productsAgent.get();
      const {
        entities: { products: normalizedProducts }
      } = normalize(products, [productSchema]);
      this.updateProductsInRegistry(normalizedProducts);
      return products;
    } catch (err) {
      throw err;
    } finally {
      this.productLoading = false;
    }
  });

  @action
  updateProductsInRegistry(products) {
    return updateItemsInRegistry(products, this.registry);
  }
}

export default new Products();
