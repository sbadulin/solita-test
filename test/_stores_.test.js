import { toJS } from "mobx";

import { Products } from "../src/stores/products";

let testInstance = null;

describe("products store", () => {
  beforeEach(() => {
    testInstance = new Products();
  });

  it("searches for products", async () => {
    const products = await testInstance.getProducts();
    expect(toJS(testInstance.registry)).toMatchSnapshot();
    expect(products.length).toBe(3);
  });

  it("searches for product with id 1", async () => {
    await testInstance.getProducts(1);
    expect(toJS(testInstance.registry[1])).toMatchSnapshot();
  });
});
