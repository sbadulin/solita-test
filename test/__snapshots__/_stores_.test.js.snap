// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`products store searches for product with id 1 1`] = `
Object {
  "description": "Perfect choice when you don’t need any corners.",
  "id": 1,
  "key": 1,
  "name": "Circle",
  "price": 999,
  "splittedDescription": "Perfect choice when you
don’t need any corners.",
}
`;

exports[`products store searches for products 1`] = `
Object {
  "1": Object {
    "description": "Perfect choice when you don’t need any corners.",
    "id": 1,
    "key": 1,
    "name": "Circle",
    "price": 999,
    "splittedDescription": "Perfect choice when you
don’t need any corners.",
  },
  "2": Object {
    "description": "For once, it’s a great idea to think inside the box.",
    "id": 2,
    "key": 2,
    "name": "Rectangle",
    "price": 899,
    "splittedDescription": "For once, it’s a great idea
to think inside the box.",
  },
  "3": Object {
    "description": "A true classic with three elegant corners.",
    "id": 3,
    "key": 3,
    "name": "Triangle",
    "price": 1009,
    "splittedDescription": "A true classic with
three elegant corners.",
  },
}
`;
